### DISCUSSION TITLE
Apple OpenID Provider

### DISCUSSION THREAD

**Discussion by MaxenceMottard** [directus/directus#9313] [Archive]
Hello, does anyone know how to make the OpenId provider `Sign in with apple` work. Here is my error :

<img width="515" alt="Capture d’écran 2021-11-01 à 17 11 36" src="https://user-images.githubusercontent.com/28975415/139703702-f649cabd-badf-475c-aa9c-42b9ecbe3d16.png">

Here is my configuration:

```env
AUTH_PROVIDERS="apple"

AUTH_APPLE_DRIVER="openid"
AUTH_APPLE_CLIENT_ID=""
AUTH_APPLE_CLIENT_SECRET= ""
AUTH_APPLE_ISSUER_URL="https://appleid.apple.com/.well-known/openid-configuration"
AUTH_APPLE_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_APPLE_DEFAULT_ROLE_ID=""
AUTH_APPLE_SCOPE="openid"
AUTH_APPLE_ICON="apple"
```

**Comment by aidenfoxx** [comment]
Apple doesn't support the `userinfo` endpoint, which we use to fetch profile information about the user. We can fetch the users email from the initial OpenID response, but any created users won't have additional profile info (first_name, last_name etc). But I guess that's better than nothing!

I'll add some code to handle instances where the provider doesn't provide a `userinfo` endpoint.

**Reply by MaxenceMottard** [reply]
Thank you for your answer. I'll try to follow the pull request on this subject.

**Reply by rijkvanzanten** [reply]
@MaxenceMottard This should do the trick: https://github.com/directus/directus/pull/9368

**Reply by biapar** [reply]
Where is the code?

**Comment by MaxenceMottard** [comment]
Hello, one last question about providers. I use `Sign In With Apple` on an iOS application. I use the native api to authenticate but after having the code on my mobile application, I am not sure if there is an equivalent request to the callback page to authenticate without going through a web page.
I hope to have been clear.

**Reply by rijkvanzanten** [reply]
Ohhh very good question. What format is that code you get from the native sign in? Is that the same as OAuth?

**Reply by MaxenceMottard** [reply]
`https://domain.com/auth/login/apple/callback?state=BCZniRuQk4fbfxQuaNai3ljQzrnF1UFtxKR5Qb0IdeQ&code=c21704d6fc05c4b6fa1fc068214afca2b.0.rvvw.9hYIVSqpYShtv6FI7l4H5Q`

Here you can see the callback called after Sign in with Apple.
I don't know what the `state` value is but from the native API on iOS I get the same `code` format.

**Reply by aidenfoxx** [reply]
State is a secure value we send in with an OAuth request, and we expect the response to contain the same value. It's required by some OAuth providers.

**Reply by MaxenceMottard** [reply]
Ok thanks for the explanation

**Reply by biapar** [reply]
> Ok thanks for the explanation

Have you solved? if yes, how?

**Reply by MaxenceMottard** [reply]
No I don't have any solution for the moment. Does anyone have anything to suggest?

**Reply by rijkvanzanten** [reply]
@MaxenceMottard I don't think there's a safe way to do this at the moment. The service that starts the OAuth flow (in this case your native app) is the one that needs to "finish" it, as there's a state w/ temporary session generated on the first request to the provider. There's no way to securely hijack the first half of the flow yourself at this point 👍🏻

**Reply by MaxenceMottard** [reply]
I'm sad 😢

**Reply by biapar** [reply]
So which are the steps to "add" Apple SSO?

**Reply by MaxenceMottard** [reply]
Sign In with Apple already works. But it's only usable on the web. So if you develop a mobile application you will have to go through webviews.

**Reply by biapar** [reply]
@MaxenceMottard like the others. Google / Facebook / Microsoft redirect people to login page and then the callback redirect to you to et the token.

**Reply by MaxenceMottard** [reply]
Yes, but it's a web approach, not a mobile approach. That's not what I'm interested in.

**Reply by biapar** [reply]
In the past I tried the Azure B2C auth and it function like that also on Mobile. 
Maybe we need to use a middle OAUTH system

**Reply by aidenfoxx** [reply]
@biapar I believe you're talking about the "explicit grant" flow, where the access token is returned in the callback? This flow is generally not recommended and is actually depreciated in OAuth 2.1.

**Reply by biapar** [reply]
> @biapar I believe you're talking about the "explicit grant" flow, where the access token is returned in the callback? This flow is generally not recommended and is actually depreciated in OAuth 2.1.

Yes it's it

**Reply by aidenfoxx** [reply]
Yeah... we don't support that and probably never will. It's being depreciated for a reason. 😅 

If you want something like that you can add a special hook to redirect with the access token, but that's at your own risk.

**Comment by eikaramba** [comment]
@rijkvanzanten pinging you, as i don't know wheter answers to an already answered discussion are pushed 

i am trying to implement apple sign in. i need both web and native apple sign in options. i used to do this in strapi [with some hacky code](https://lith0x.medium.com/implement-apple-sign-in-with-strapi-a9f0ff4b6417) and the help of https://github.com/renarsvilnis/apple-sign-in-rest

from this discussion it is not clear how i can support both options in directus. supabase for example has a [guide](https://supabase.com/docs/guides/auth/auth-apple) which refers to manually generating the client_secret, but this is only valid for 6 monts. the nodejs library can generate this for you, but how can i integrate this into the flow?

```js
async (code, origin) => {
    // read your origin to define the clientId. origin is basically defined based on whether the call is coming from the browser or from native side. e.g. looking at referer header or query params...
    const clientId = origin ? "xx.xxxxx.weblogin" : "xx.xxxx.nativelogin"; // change this with your own value

    const appleSignIn = new AppleSignIn({
      /**
       * This clientId depends on that login "flow" you trying to create:
       *   - "web login" - this is the "serviceId"
       *   - "ios login" - this is the app "bundleId", choose only this if you trying to
       *                   verify user that has signed into using the native iOS way
       *
       */
      clientId: clientId,
      teamId: "xxxxxxxxxxx",
      keyIdentifier: "xxxxxxxxxxxx",
      privateKeyPath: __dirname + "/../config/AuthKey_xxxxxxxx.p8" // the name of the AuthKey previously copy/paste under config folder
    });

    const clientSecret = appleSignIn.createClientSecret({
      expirationDuration: 20 * 60 // define your own expiration client secret
    });
    const tokenResponse = await appleSignIn.getAuthorizationToken(
      clientSecret,
      code,
      {}
    );
    return await appleSignIn.verifyIdToken(tokenResponse.id_token, {});
  }
  ```

**Reply by eikaramba** [reply]
to extend further and provide some insights how it is currently working for me on mobile (again with strapi, which i want to ditch in favor of directus *G*)

there you call a native method on ios which handles the whole ui flow and gives back the  credentials.
https://docs.expo.dev/versions/latest/sdk/apple-authentication/#appleauthenticationcredential (i'm not using expo, but it has a nice overview of the native methods etc.)
```js
function openOauth(provider){
        if(provider=="apple" && isSignInWithAppleSupported()){
                signInWithApple({
                    scopes: ["EMAIL", "FULLNAME"]
                })
                .then(async (result) => {
                    //pass result.credential.authorizationCode and state to backend and redirect user to main page
                })
```

i would now _assume_ i need to pass the **authorizationCode and state** to directus's /auth/login/apple/callback, which would evidently create the user in the db and gives back the access_token&refres_token. unfortunately in this case [the cookie](https://github.com/directus/directus/blob/main/api/src/auth/drivers/openid.ts#L252) is not set however...

**Reply by eikaramba** [reply]
i think i might have an idea at least for the first part of the problem (calling apple endpoint with generated client_secret. i just implement the whole authentication flow myself ;) redirecting user to correct url and everything. only at the very end i pass the authorizationCode and state to directus (which before was never invoked in this flow). if i also set the cookie for directus this might even work */endorphin level rises/*

will get back here. hopefully this helps others too.

**Reply by aidenfoxx** [reply]
Directus uses PKCE so just `code` and `state` won't be enough. It also require a `code_verifier` which is currently stored in a JWT cookie in the users session. Although I may be misunderstanding what you're trying to do.

**Reply by eikaramba** [reply]
no you are right, i think i could just set the JWT cookie myself in my example. i sure hope this works :) if you have a different idea i am all ears open

Web:
```
[
client(create JWT cookie)
-> Apple authorization page
-> client (createClientSecret)
]
-> directus /auth/login/apple/callback
```

ios:
```
[
client(remember code_verifier)
-> Apple authorization page
-> client (createClientSecret) set headers.cookie with JWT cookie manually
]
-> directus /auth/login/apple/callback
```

`[ ]` = implement myself as custom endpoints


why do all this?
- no need to recreate client_secret every 6 months, it is automatically generated by forementioned nodejs library
- i don't know how to start the process without this on mobile

**Reply by aidenfoxx** [reply]
For IOS native, why wouldn't you just use a webview with something like "/auth/login/apple?redirect=<success_url>", and read the users Directus cookie when they hit the "<success_url>"? 🤔

**Reply by eikaramba** [reply]
i might at some point. i thought it is required to use the native sign in method, but yeah... it seems not. gonna have to think about it. and for the client_secret i might just setup a cron job or so ;)

**Reply by MaxenceMottard** [reply]
> For IOS native, why wouldn't you just use a webview with something like "/auth/login/apple?redirect=<success_url>", and read the users Directus cookie when they hit the "<success_url>"? 🤔 

Because it's ugly. And it's a bad user experience for an iPhone user.

**Reply by aidenfoxx** [reply]
> Because it's ugly. And it's a bad user experience for an iPhone user.

And frankensteining 2 seperate oauth flows together isn't ugly? There is no nice way around this 😂

**Reply by eikaramba** [reply]
some update from me.
i was able to now make the normal apple web flow and the native apple flow work.

basically i have 2 oauth providers, one for apple web and apple native in the config file
```
AUTH_APPLE_DRIVER="openid"
AUTH_APPLE_CLIENT_ID="com.project.xxxx"
AUTH_APPLE_CLIENT_SECRET=xxx
AUTH_APPLE_ISSUER_URL="https://appleid.apple.com/.well-known/openid-configuration"
AUTH_APPLE_ALLOW_PUBLIC_REGISTRATION="true"
AUTH_APPLE_IDENTIFIER_KEY="sub"
AUTH_APPLE_DEFAULT_ROLE_ID="xxxx"
AUTH_APPLE_SCOPE="openid"
AUTH_APPLE_ICON="apple"
```
copy this and change the client_id. for apple you always have a service credential for the web login (with its own client id) and the app bundle id for the native login as the client_id

now i created a cron job to regularly update the secret, as it is only valid for max 6 months. (you will need to add some env variables, see the code)
```js
const fs = require("fs");
const os = require("os");
const {AppleSignIn} = require("apple-sign-in-rest");

async function generateSecret(env,forWeb) {
  // read your origin to define the clientId
  const clientId = forWeb ? env.AUTH_APPLE_CLIENT_ID : env.AUTH_APPLENATIVE_CLIENT_ID; // change this with your own value

  const appleSignIn = new AppleSignIn({
    clientId: clientId,
    teamId: env.APPLE_TEAM_ID,
    keyIdentifier: env.APPLE_KEY_IDENTIFIER,
    privateKeyPath: __dirname + "/../../../AuthKey_"+env.APPLE_KEY_IDENTIFIER+".p8"
  });
  //make secret valid for 90 days. max 180 according to apple
  const clientSecret = appleSignIn.createClientSecret({
    expirationDuration: 86400 * 60
  });
  return clientSecret;
}

function setEnvValue(key, value) {

    // read file from hdd & split if from a linebreak to a array
    const ENV_VARS = fs.readFileSync(process.env.CONFIG_PATH, "utf8").split('\n'); //or use os.EOL

    // find the env we want based on the key
    const target = ENV_VARS.indexOf(ENV_VARS.find((line) => {
        return line.match(new RegExp(key));
    }));

    // replace the key/value with the new value
    ENV_VARS.splice(target, 1, `${key}=${value}`);

    // write everything back to the file system
    fs.writeFileSync(process.env.CONFIG_PATH, ENV_VARS.join('\n'));
}

module.exports = function registerHooks({ schedule },{env}) {
	schedule('0 3 1 * *', async () => {
		setEnvValue("AUTH_APPLE_CLIENT_SECRET", await generateSecret(env,true));
		setEnvValue("AUTH_APPLENATIVE_CLIENT_SECRET", await generateSecret(env,false));
		
		//needed to refresh environment variables. server will respawn by pm2
		setTimeout(() => {
			process.exit(0);
		}, 2000);
	});
};
```

then login via web follows the same logic than any oauth provider. i had some big problems in regards to a bug/feature in FF/safari ios which prevents reading the cookie if a redirect is involved. basically reading the cookie is not possible when directus redirects to a url. it only works if the redirect page refreshes itself again. i did this with the meta html attribute, then the browser was sending the cookie correctly.  you can read more about this in my little [direcuts missing documentation](https://fork-appeal-054.notion.site/Directus-Missing-Manual-07cfbb75414b4fcb9401f4325701d61e) page, which i started :) 

the native flow however was tricky. we need to completely bypass the directus authentication, because you don't have this code_verifier. also there is no authorize web link in the beginning. thus i recreated the flow (getting token, creating user, creating session etc.) in a custom endpoint. i will not release this for now, it should be secure, but i want to double check. just know that this is possible and you can basically recreate the authentication service here.
this means the flow is like this
native ios requests code -> sends code to backend endpoint -> creates or update user with access_token and refresh_token and returns this so that native app can store it.

**Reply by kkoskenvirta** [reply]
> some update from me. i was able to now make the normal apple web flow and the native apple flow work.
> 
> basically i have 2 oauth providers, one for apple web and apple native in the config file
> 
> ```
> AUTH_APPLE_DRIVER="openid"
> AUTH_APPLE_CLIENT_ID="com.project.xxxx"
> AUTH_APPLE_CLIENT_SECRET=xxx
> AUTH_APPLE_ISSUER_URL="https://appleid.apple.com/.well-known/openid-configuration"
> AUTH_APPLE_ALLOW_PUBLIC_REGISTRATION="true"
> AUTH_APPLE_IDENTIFIER_KEY="sub"
> AUTH_APPLE_DEFAULT_ROLE_ID="xxxx"
> AUTH_APPLE_SCOPE="openid"
> AUTH_APPLE_ICON="apple"
> ```
> 
> copy this and change the client_id. for apple you always have a service credential for the web login (with its own client id) and the app bundle id for the native login as the client_id
> 
> now i created a cron job to regularly update the secret, as it is only valid for max 6 months. (you will need to add some env variables, see the code)
> 
> ```js
> const fs = require("fs");
> const os = require("os");
> const {AppleSignIn} = require("apple-sign-in-rest");
> 
> async function generateSecret(env,forWeb) {
>   // read your origin to define the clientId
>   const clientId = forWeb ? env.AUTH_APPLE_CLIENT_ID : env.AUTH_APPLENATIVE_CLIENT_ID; // change this with your own value
> 
>   const appleSignIn = new AppleSignIn({
>     clientId: clientId,
>     teamId: env.APPLE_TEAM_ID,
>     keyIdentifier: env.APPLE_KEY_IDENTIFIER,
>     privateKeyPath: __dirname + "/../../../AuthKey_"+env.APPLE_KEY_IDENTIFIER+".p8"
>   });
>   //make secret valid for 90 days. max 180 according to apple
>   const clientSecret = appleSignIn.createClientSecret({
>     expirationDuration: 86400 * 60
>   });
>   return clientSecret;
> }
> 
> function setEnvValue(key, value) {
> 
>     // read file from hdd & split if from a linebreak to a array
>     const ENV_VARS = fs.readFileSync(process.env.CONFIG_PATH, "utf8").split('\n'); //or use os.EOL
> 
>     // find the env we want based on the key
>     const target = ENV_VARS.indexOf(ENV_VARS.find((line) => {
>         return line.match(new RegExp(key));
>     }));
> 
>     // replace the key/value with the new value
>     ENV_VARS.splice(target, 1, `${key}=${value}`);
> 
>     // write everything back to the file system
>     fs.writeFileSync(process.env.CONFIG_PATH, ENV_VARS.join('\n'));
> }
> 
> module.exports = function registerHooks({ schedule },{env}) {
> 	schedule('0 3 1 * *', async () => {
> 		setEnvValue("AUTH_APPLE_CLIENT_SECRET", await generateSecret(env,true));
> 		setEnvValue("AUTH_APPLENATIVE_CLIENT_SECRET", await generateSecret(env,false));
> 		
> 		//needed to refresh environment variables. server will respawn by pm2
> 		setTimeout(() => {
> 			process.exit(0);
> 		}, 2000);
> 	});
> };
> ```
> 
> then login via web follows the same logic than any oauth provider. i had some big problems in regards to a bug/feature in FF/safari ios which prevents reading the cookie if a redirect is involved. basically reading the cookie is not possible when directus redirects to a url. it only works if the redirect page refreshes itself again. i did this with the meta html attribute, then the browser was sending the cookie correctly. you can read more about this in my little [direcuts missing documentation](https://fork-appeal-054.notion.site/Directus-Missing-Manual-07cfbb75414b4fcb9401f4325701d61e) page, which i started :)
> 
> the native flow however was tricky. we need to completely bypass the directus authentication, because you don't have this code_verifier. also there is no authorize web link in the beginning. thus i recreated the flow (getting token, creating user, creating session etc.) in a custom endpoint. i will not release this for now, it should be secure, but i want to double check. just know that this is possible and you can basically recreate the authentication service here. this means the flow is like this native ios requests code -> sends code to backend endpoint -> creates or update user with access_token and refresh_token and returns this so that native app can store it.

Hey! Did you complete the custom endpoint for the native flow? Would you be interested in sharing this approach as a starting point to learn how this should be implemented? This would be highly appreciated 👍  @eikaramba

**Reply by eikaramba** [reply]
yes i did, just copying the code here: 

```js
const {BaseException} = require('@directus/shared/exceptions');
const {AppleSignIn} = require("apple-sign-in-rest");
const jwt = require('jsonwebtoken');
const {nanoid} = require('nanoid');
const ms = require('ms');

module.exports = function(router,{ services, exceptions, logger,env }){
  const { ItemsService } = services;
	const { ServiceUnavailableException,RouteNotFoundException,ForbiddenException   } = exceptions;

	router.get('/applenative/callback', async (req, res, next) => {
    const usersServiceAdmin = new ItemsService('directus_users', { schema: req.schema });
    const sessionServiceAdmin = new ItemsService('directus_sessions', { schema: req.schema });

    console.log("apple claim",req.query)

    if(!req.query.code) throw new BaseException("code missing",400);


    // // read your origin to define the clientId
    const clientId = "xx.yyy.ww"; // change this with your own value

    try {
      const appleSignIn = new AppleSignIn({
        /**
         * This clientId depends on that login "flow" you trying to create:
         *   - "web login" - this is the "serviceId"
         *   - "ios login" - this is the app "bundleId", choose only this if you trying to
         *                   verify user that has signed into using the native iOS way
         *
         */
        clientId: clientId,
        teamId: "xxxxxxx",
        keyIdentifier: "YYYYYYYYYY",
        privateKeyPath: __dirname + "/../../../AuthKey_xxxxxxxxxxx.p8" // the name of the AuthKey previously copy/paste under config folder
      });
  
      // console.log("clientId",clientId)
      // const clientSecret = appleSignIn.createClientSecret({
      //   expirationDuration: 20 * 60 // define your own expiration client secret
      // });
  
      const tokenResponse = await appleSignIn.getAuthorizationToken(
        env.AUTH_APPLENATIVE_CLIENT_SECRET,
        req.query.code,
        {}
      );
      console.log(tokenResponse);
      if(!tokenResponse) throw new BaseException("invalid token response",401);
      

      let claim = await appleSignIn.verifyIdToken(req.query.identityToken, {
        subject: req.query.user,
      });
      if(!claim.sub){
        claim = await appleSignIn.verifyIdToken(tokenResponse.id_token, {
          subject: req.query.user,
        });
      }


      console.log(claim);
      if(!claim) throw new BaseException("invalid claim",401);

      let user = await usersServiceAdmin.readByQuery({fields:["id"],limit:1,filter:{
        external_identifier: {_eq:claim.sub}
      }});
      user = user[0];

      if (user) {
        // Update user refreshToken if provided
        if (tokenResponse.refresh_token) {
          await usersServiceAdmin.updateOne(user.id, {
            auth_data: JSON.stringify({ refreshToken: tokenResponse.refresh_token }),
          });
        }
      }else{
        user = await usersServiceAdmin.createOne({
          provider: "applenative",
          first_name: "",
          last_name: "",
          email: claim.email,
          external_identifier: claim.sub,
          role: env.AUTH_APPLENATIVE_DEFAULT_ROLE_ID,
          auth_data: tokenResponse.refresh_token && JSON.stringify({ refreshToken: tokenResponse.refresh_token }),
        });
        user = {id:user};
      }
      console.log("user",user);

      //TODO: app_access should come from role setting
      const access_token = jwt.sign({
        id: user.id,
        role: env.AUTH_APPLENATIVE_DEFAULT_ROLE_ID,
        app_access:false,
        admin_access:false,
      }, env.SECRET, {
        expiresIn: env.ACCESS_TOKEN_TTL,
        issuer: 'directus',
      });
      const refresh_token = nanoid(64);
      const refreshTokenExpiration = new Date(Date.now() + ms(env.REFRESH_TOKEN_TTL));


      await sessionServiceAdmin.createOne({
        token: refresh_token,
        user: user.id,
        expires: refreshTokenExpiration
      });
      //TODO: Remove debug statements
      console.log("final response",{
        access_token,
        refresh_token,
        expires: ms(env.ACCESS_TOKEN_TTL)
      });

      res.send( {
        access_token,
        refresh_token,
        expires: ms(env.ACCESS_TOKEN_TTL)
      });






    } catch (err) {
      console.log(err);
    }
  });

}
```

please test it yourself and report back if you want issues or things to improve. it is tested on 9.17.0 or so, should also work on latest but haven't tested it yet. i basically copied the original auth code and adapted it slightly to use the token from apple

**Reply by kkoskenvirta** [reply]
Thank you so much @eikaramba! This saved me multiple days of work 🙏  The code worked for me with minor changes to some package importing formats. Im not much of an expert with the directus extensions so this really opened the world of possibilities for me.

**Reply by eikaramba** [reply]
Glad that it works for you. Yes i spent wayyyy tooo long on the whole apple native authentication.

**Reply by NilsBaumgartner1994** [reply]
> > some update from me. i was able to now make the normal apple web flow and the native apple flow work.
> > basically i have 2 oauth providers, one for apple web and apple native in the config file
> > ```
> > AUTH_APPLE_DRIVER="openid"
> > AUTH_APPLE_CLIENT_ID="com.project.xxxx"
> > AUTH_APPLE_CLIENT_SECRET=xxx
> > AUTH_APPLE_ISSUER_URL="https://appleid.apple.com/.well-known/openid-configuration"
> > AUTH_APPLE_ALLOW_PUBLIC_REGISTRATION="true"
> > AUTH_APPLE_IDENTIFIER_KEY="sub"
> > AUTH_APPLE_DEFAULT_ROLE_ID="xxxx"
> > AUTH_APPLE_SCOPE="openid"
> > AUTH_APPLE_ICON="apple"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > copy this and change the client_id. for apple you always have a service credential for the web login (with its own client id) and the app bundle id for the native login as the client_id
> > now i created a cron job to regularly update the secret, as it is only valid for max 6 months. (you will need to add some env variables, see the code)
> > ```js
> > const fs = require("fs");
> > const os = require("os");
> > const {AppleSignIn} = require("apple-sign-in-rest");
> > 
> > async function generateSecret(env,forWeb) {
> >   // read your origin to define the clientId
> >   const clientId = forWeb ? env.AUTH_APPLE_CLIENT_ID : env.AUTH_APPLENATIVE_CLIENT_ID; // change this with your own value
> > 
> >   const appleSignIn = new AppleSignIn({
> >     clientId: clientId,
> >     teamId: env.APPLE_TEAM_ID,
> >     keyIdentifier: env.APPLE_KEY_IDENTIFIER,
> >     privateKeyPath: __dirname + "/../../../AuthKey_"+env.APPLE_KEY_IDENTIFIER+".p8"
> >   });
> >   //make secret valid for 90 days. max 180 according to apple
> >   const clientSecret = appleSignIn.createClientSecret({
> >     expirationDuration: 86400 * 60
> >   });
> >   return clientSecret;
> > }
> > 
> > function setEnvValue(key, value) {
> > 
> >     // read file from hdd & split if from a linebreak to a array
> >     const ENV_VARS = fs.readFileSync(process.env.CONFIG_PATH, "utf8").split('\n'); //or use os.EOL
> > 
> >     // find the env we want based on the key
> >     const target = ENV_VARS.indexOf(ENV_VARS.find((line) => {
> >         return line.match(new RegExp(key));
> >     }));
> > 
> >     // replace the key/value with the new value
> >     ENV_VARS.splice(target, 1, `${key}=${value}`);
> > 
> >     // write everything back to the file system
> >     fs.writeFileSync(process.env.CONFIG_PATH, ENV_VARS.join('\n'));
> > }
> > 
> > module.exports = function registerHooks({ schedule },{env}) {
> > 	schedule('0 3 1 * *', async () => {
> > 		setEnvValue("AUTH_APPLE_CLIENT_SECRET", await generateSecret(env,true));
> > 		setEnvValue("AUTH_APPLENATIVE_CLIENT_SECRET", await generateSecret(env,false));
> > 		
> > 		//needed to refresh environment variables. server will respawn by pm2
> > 		setTimeout(() => {
> > 			process.exit(0);
> > 		}, 2000);
> > 	});
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > then login via web follows the same logic than any oauth provider. i had some big problems in regards to a bug/feature in FF/safari ios which prevents reading the cookie if a redirect is involved. basically reading the cookie is not possible when directus redirects to a url. it only works if the redirect page refreshes itself again. i did this with the meta html attribute, then the browser was sending the cookie correctly. you can read more about this in my little [direcuts missing documentation](https://fork-appeal-054.notion.site/Directus-Missing-Manual-07cfbb75414b4fcb9401f4325701d61e) page, which i started :)
> > the native flow however was tricky. we need to completely bypass the directus authentication, because you don't have this code_verifier. also there is no authorize web link in the beginning. thus i recreated the flow (getting token, creating user, creating session etc.) in a custom endpoint. i will not release this for now, it should be secure, but i want to double check. just know that this is possible and you can basically recreate the authentication service here. this means the flow is like this native ios requests code -> sends code to backend endpoint -> creates or update user with access_token and refresh_token and returns this so that native app can store it.
> 
> Hey! Did you complete the custom endpoint for the native flow? Would you be interested in sharing this approach as a starting point to learn how this should be implemented? This would be highly appreciated 👍 @eikaramba

I have the problem, that this mostly works and i get a new secret generated, but:
- directus only accepts it, when i completly restart the docker. It seems that the env variable are used somewhere earlier.

Any ideas on how to fix this? I dont want manually to restart the server.
